artemis-core-client/src/main/java/org/apache/activemq/artemis/core/message/impl/CoreMessage.java

```diff
@@ -52,6 +52,62 @@
  *  consumers */
 public class CoreMessage extends RefCountMessage implements ICoreMessage {
 
+   private static final class CoreTypedProperties extends TypedProperties {
+
+      // We use properties to establish routing context on clustering.
+      // However if the client resends the message after receiving, it needs to be removed
+      private static final Predicate<SimpleString> INTERNAL_PROPERTY_NAMES_CLEANUP_FILTER =
+         name -> (name.startsWith(Message.HDR_ROUTE_TO_IDS) && !name.equals(Message.HDR_ROUTE_TO_IDS)) ||
+            (name.startsWith(Message.HDR_ROUTE_TO_ACK_IDS) && !name.equals(Message.HDR_ROUTE_TO_ACK_IDS));
+      private static final SimpleString AMQ_PROPNAME = new SimpleString("_AMQ_");
+      private boolean internalProperties;
+
+      CoreTypedProperties() {
+         super();
+         internalProperties = false;
+      }
+
+      private CoreTypedProperties(TypedProperties other) {
+         super(other);
+         if (other instanceof CoreTypedProperties) {
+            internalProperties = ((CoreTypedProperties) other).internalProperties;
+         } else {
+            internalProperties = other.containsProperty(name -> name.startsWith(AMQ_PROPNAME));
+         }
+      }
+
+      @Override
+      protected synchronized void doPutValue(SimpleString key, TypedProperties.PropertyValue value) {
+         if (!internalProperties && key.startsWith(AMQ_PROPNAME)) {
+            internalProperties = true;
+         }
+         super.doPutValue(key, value);
+      }
+
+      public synchronized boolean cleanupInternalProperties() {
+         if (!internalProperties) {
+            return false;
+         }
+         return removeProperty(INTERNAL_PROPERTY_NAMES_CLEANUP_FILTER);
+      }
+
+      public static boolean cleanupInternalProperties(TypedProperties typedProperties) {
+         if (typedProperties == null) {
+            return false;
+         }
+         if (typedProperties instanceof CoreTypedProperties) {
+            return ((CoreTypedProperties) typedProperties).cleanupInternalProperties();
+         }
+         return typedProperties.removeProperty(INTERNAL_PROPERTY_NAMES_CLEANUP_FILTER);
+      }
+
+      public static CoreTypedProperties copyOf(TypedProperties typedProperties) {
+         synchronized (typedProperties) {
```

Comment:
this shouldnt be needed, as the constructor of CoreTypedProperties that takes TypedProperties should really be the thing that ensure syncronicitity

Reply:
You can't put a synchronized before the call to `super(other)`, that's why I have used a factory: if I will drop `CoreTypedProperties` all of this will disappear right?

Reply:
Of course

Reply:
Nice: I have removed subclass of `TypedProperties` and cleaned up the code: it seems much better now. The only thing that bother me is that I would like to avoid at all to perform that `cleanupInternalProperties`, given that most of time it won't find anything; i just need to understand what's its purpose and how to know in advance when do it is a waste of time.

Reply:
+1 agreed this would need to be resolved as currently thats why the flag exist to avoid the iterating over the properties.

Reply:
Yep, I was expecting it to be, but reality was that even with that flag it would fail and will iterate through the properties: the flag was only checking if *any* internal properties is being put, while the `cleanupInternlProperties` aim to cleanup just specific internal ones.

Reply:
So it being internal, means where this gets put MUST be somewhere in Artemis code base, maybe a small nullable boolean flag on the CoreMessage could be introduced that is set after by that code when these properties are added.